src/main/java/org/jenkinsci/plugins/workflow/cps/CpsBodyExecution.java

```diff
@@ -247,13 +254,20 @@ public boolean cancel(final CauseOfInterruption... causes) {
             t.getExecution().runInCpsVmThread(new FutureCallback<CpsThreadGroup>() {
                 @Override
                 public void onSuccess(CpsThreadGroup g) {
+                    if (g.threads.size() == 1) {
```

Comment:
I can't optimize this with the block lookup cache since the body isn't necessarily enclosing (boo), but at least can make it faster in the common case.